Yoast/wordpress-seo
# Fix notice dismissal bug

## Context


* In the new General page, admin-wide notices will reappear once dismissed when switching from a section to another.

## Summary


This PR can be summarized in the following changelog entry:

* Fixes a bug where dismissed notices would re-appear when switching from a section to another of the Yoast general page.

## Relevant technical choices:

* I used the `MutationObserver` interface to react to a notice being hidden when its close button is pressed. By doing so the original Javascript logic in each notice's integration is still executed, but when the notice `display` attribute changes to `none` the observer's callback updates also `resolvedNotices` in the Redux store.
* The `useEffect` which sets up the notices observers in the `App` component has `pathname` as dependency because the observers need to be re-created each time the user browse back from the `First-time configuration` section: this is because in that section the notices are not rendered and, from what I understood, the observers are lost.
* I renamed the slice we use to store the resolved notices from `firstTimeConfiguration` to `yoastAdminNotices` because it is not really tied to the First-time configuration, but it is tied to the notices we show in the admin pages.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Preliminary steps
* Install Yoast SEO Premium but don't activate it
* Start the First-time configuration but don't complete it
* Use the Yoast Test Helper to reset the options

#### Test the fix
* Go to `Yoast SEO` -> `General`
* Verify you see the following two dismissible notices:
<img width="1036" alt="Screenshot 2024-11-14 at 00 46 03" src="https://github.com/user-attachments/assets/cedc988c-30c2-413c-baff-3dd0d22c431e">

* Dismiss one of the two notices
* Go to `First-time configuration` and then back to `Alert center`
* Verify the notice is still dismissed
* Perform the same check with the other notice
* Use the Yoast Test Helper to reset the options
* Go to `Yoast SEO` -> `General`
* Dismiss both notices
* Go to `First-time configuration` and then back to `Alert center`
* Verify the notices are still dismissed
* Use the Yoast Test Helper to reset the options
* Dismiss the `Activate Yoast SEO Premium!` notice you see in the Yoast Test Helper page
* Go to `Yoast SEO` -> `General`
* Verify the notice is dismissed

#### Relevant test scenarios


### Test instructions for QA when the code is in the RC



## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

* N/A

## UI changes


## Other environments


## Documentation


## Quality assurance


## Innovation


Fixes [#316](https://github.com/Yoast/reserved-tasks/issues/316)
